### DISCUSSION TITLE
remote sensors: get channel number that succeeded (ESPnow)

### DISCUSSION THREAD

**Discussion by homonto** [HomeSpan/HomeSpan#438] [Ideas]
Hi, I implemented the fixed channel as per your description and it works like a charm.
On top of that, I implemented captive portal where you can set up wifi credentials together with fixed channel
However sometimes you need to change channel on router - so in my implementation I need to push remote device into captive portal and change the channel again.
Actually what I did is:
1- when wake up, check which channel is configured in Preferences
2- use this channel to send data
3- if FAILED, delete the channel from Preferences 
4- since now on, use all channels (1-14 probably, mine is 8 so I am not sure if your program goes to 13 or 14)
but now there is an issue: from 2ms connection time I am at almost 1s.

My idea is: your program knows which channel SUCCEEDED as I can see in LOG 2, can I get somehow the number of the channel that succeeded?  (but not from the Serial terminal - rather in the program itself). Then, once succeeded and I know which channel it was, I could save in Preferences again this channel and since now on, I would be back at 2ms on ESPnow rather than 1s.

Thank you for your help.

**Comment by homonto** [comment]
OK, I found a solution - dirty but works.
1- send over 1 channel that is saved in preferences
2- if it fails, go through the loop from channel 1 to 13 and send again
3- in case of first success save the channel in preferences (channel number from the loop "i")
so in the worse case scenario 1 time sensor has to go through all channels but next time it uses the saved channel

**Reply by blaisebarrette** [reply]
Very interesting!
I am working on a project involving espNow.
Would you share your code? Or the part you described for saving the working Chanel?

Thanks!

**Reply by homonto** [reply]
hi @blaisebarrette you are very welcome of course - here you are:
https://github.com/homonto/HomeKit_on_HomeSpan

and below is the part of saving time by using 1 channel:

```

  preferences.begin("wifi-secrets", true); // read only
  channel = preferences.getString("rec_channel", "0");
  sleeptime_s_str = preferences.getString("rec_sleep_s", "0");
  preferences.end();
  int wifi_channel = channel.toInt();
  sleeptime_s = sleeptime_s_str.toInt();
// check channel
  bool valid_wifi_channel = false;
  if ((wifi_channel > 0) and (wifi_channel < 14))
  {
    valid_wifi_channel = true;
    #ifdef DEBUG
      Serial.printf("[%s]: Found saved WiFi channel=%d - will use it for ESPnow at first\n",__func__,wifi_channel);
    #endif
  } else 
  {
    // #ifdef DEBUG
      Serial.printf("[%s]: Found saved WiFi channel=%d is incorrect - will NOT use it for ESPnow!\n",__func__,wifi_channel);
    // #endif
  }

  // sending data to bridge
  #ifdef DEBUG
    Serial.printf("[%s]: Sending data to bridge...\n",__func__);
  #endif
  mainDevice=new SpanPoint(BRIDGE,sizeof(myData),0);    // create a SpanPoint with send size=sizeof(myData) and receive size=0
  // is ESPnow channel fixed?
/*
if valid channel:
{
  send, 
  if successfull 
  {
    bye
  } else 
  {
    repeat sending for all channels from 1-14
    if successful
    {
      save
      bye
    } else
    {
      bye
    }
  }
} else // no channel provided 
{
  sending for all channels from 1-14
  if successful
  {
    save
    bye
  } else
  {
    bye
  }
}
*/

  unsigned long start_send_data = millis();
  unsigned long end_send_data = 0;
  bool send_successful = false;
  if (valid_wifi_channel)
  {
    #ifdef DEBUG
      Serial.printf("[%s]: ESPnow configured to use only one channel=%d\n",__func__,wifi_channel);
    #endif
    SpanPoint::setChannelMask(1<<wifi_channel);
    if (mainDevice->send(&myData))      // first sending, either over 1 channel or over all channels
    {
      end_send_data = millis();
      Serial.printf("[%s]: Sending data to bridge - SUCCESSFULL, it took %ums\n",__func__,(end_send_data - start_send_data));
      send_successful = true;
    } else
    {
      Serial.printf("[%s]: Sending data to bridge - this time over all WiFi channels...\n",__func__);
      start_send_data = millis();
      for (int i=1;i<14;i++)
      {
        SpanPoint::setChannelMask(1<<i);
        if (mainDevice->send(&myData))
        {
          end_send_data = millis();
          send_successful = true;

          Serial.printf("[%s]: Sending data to bridge - SUCCESSFULL, it took %ums\n",__func__,(end_send_data - start_send_data));
          Serial.printf("[%s]: Setting and saving new WiFi channel (CHANNEL=%d) in preferences\n",__func__,i);

          preferences.begin("wifi-secrets", false);     // read/write
          preferences.putString("rec_channel",String(i));
          preferences.end();
          delay(10);
          break;
        } 
      }
      if (!send_successful)
      {
        end_send_data = millis();
        Serial.printf("[%s]: Sending data to bridge - FAILED AGAIN (on ALL CHANNELS), it took %ums\n",__func__,(end_send_data - start_send_data));
      }
    }
  } else 
  {
    Serial.printf("[%s]: Sending data to bridge over all WiFi channels...\n",__func__);
    start_send_data = millis();
    for (int i=1;i<14;i++)
    {
      SpanPoint::setChannelMask(1<<i);
      if (mainDevice->send(&myData))
      {
        end_send_data = millis();
        send_successful = true;

        Serial.printf("[%s]: Sending data to bridge - SUCCESSFULL, it took %ums\n",__func__,(end_send_data - start_send_data));
        Serial.printf("[%s]: Setting and saving new WiFi channel in preferences\n",__func__);

        preferences.begin("wifi-secrets", false);     // read/write
        preferences.putString("rec_channel",String(i));
        preferences.end();
        delay(10);
        break;
      } 
    }
    if (!send_successful)
    {
      end_send_data = millis();
      Serial.printf("[%s]: Sending data to bridge - FAILED AGAIN (on ALL CHANNELS), it took %ums\n",__func__,(end_send_data - start_send_data));
    }
  }
  // sending data to bridge END

```

**Reply by blaisebarrette** [reply]
Wow! thank you sooo much!
I am currently working on the PCB this will be precious when I start working on the code! Thanks again!

**Reply by homonto** [reply]
I have done some other projects with ESPnow as well including PCB - you can see here:  https://github.com/homonto/ESPnow-MQTT-HomeAssistant-Gateway

**Comment by HomeSpan** [comment]
@homonto, I added a feature to SpanPoint that automatically stores the last channel used in NVS so that upon reboot this channel is automatically selected.  This prevents HomeSpan from performing a search starting each time at channel 1.  However, if upon reboot the last channel used is no longer valid (perhaps the Hub device started using a new channel), HomeSpan will search for new channels to use as it did before.  Once it finds a channel that succeeds, that channel is now stored in NVS for use upon reboot.

Also, HomeSpan checks to make sure whatever channel is stored is allowed by any channel mask you set.  If you change the channel mask and the stored channel is no longer allowed, it will not be used upon reboot.  Instead, the first allowed channel according to the mask will be used.

The code for this new feature has been published in the "dev" branch and will be part of the next full release.  Please feel free to give it a try and let me know how it works (should make your sketch simpler, as you won't need to store this info yourself).

**Reply by homonto** [reply]
super nice
will the code be backward compatible?
I am asking because I implemented this already in my code and I am wondering if I recompile my previous sketch will kill the code or not

**Reply by HomeSpan** [reply]
The changes should (hopefully) not impact any other code.
